[DR-2793] Add endpoint to sync DUOS Firecloud group members

okotsopoulos opened the pull request:
https://broadworkbench.atlassian.net/browse/DR-2793
Related ticket to expand Sam configuration: https://broadworkbench.atlassian.net/browse/DR-2839

We currently allow snapshot stewards to link their snapshots to DUOS datasets, which adds a Firecloud group for the DUOS dataset's authorized users as a snapshot reader.  But the groups are empty containers, only containing the TDR SA as an admin.

This PR adds support for updating the members of a DUOS Firecloud group with the linked DUOS dataset's approved users.

I broke this PR up into meaningful commits if reviewers would prefer to walk through them, and have my [dev environment](https://jade-ok.datarepo-dev.broadinstitute.org/swagger-ui.html#/duos) up-to-date.

**New Endpoints**
[retrieveDuosFirecloudGroup](https://jade-ok.datarepo-dev.broadinstitute.org/swagger-ui.html#/duos/retrieveDuosFirecloudGroup) - a logged-in user can retrieve metadata on a TDR-managed DUOS Firecloud group for a specified DUOS ID.
- Throws a 404 if no DUOS dataset with that ID has been registered in TDR

[retrieveDuosFirecloudGroups](https://jade-ok.datarepo-dev.broadinstitute.org/swagger-ui.html#/duos/retrieveDuosFirecloudGroups) - a logged-in user can retrieve metadata on all TDR-managed DUOS Firecloud groups.

[syncDuosDatasetAuthorizedUsers](https://jade-ok.datarepo-dev.broadinstitute.org/swagger-ui.html#/duos/syncDuosDatasetAuthorizedUsers) -  a user holding the `sync_duos_users` on the `datarepo` resource can sync the DUOS dataset's authorized users as members of their corresponding TDR-managed Firecloud group.
- Throws a 404 if no DUOS dataset with that ID has been registered in TDR
- Fetches the DUOS dataset's authorized users.
  - If users are fetched successfully, use them to overwrite the Firecloud group's members.
  - If DUOS tells us that the dataset does not exist or the ID is invalid, empty out the Firecloud group's members. 
  - Allow other exceptions to bubble up and treat as an unsuccessful sync.
- Update DUOS Firecloud group record's `last_synced_time`.  Concurrent modifications are unlikely but possible, and would result in a thrown 409.

**Other Notable Changes**
- DUOS is a non-critical dependency of TDR, and is now included it in our [service status check](https://jade-ok.datarepo-dev.broadinstitute.org/swagger-ui.html#/unauthenticated/serviceStatus).
- Introduced `GoogleCredentialsService` for obtaining default credentials and tokens, for easier mocking in tests. It's used in `IamService` following a refactor as well as `DuosClient`.

**Coming Soon**
- When creating a Firecloud group, also sync its contents.
- Add a DUOS endpoint to force a sync of all TDR-registered Firecloud group members.

okotsopoulos committed:
Add IamService.overwriteGroupPolicyEmails

Add new IamRole enum value MEMBER - used for managing Firecloud managed group members.
Added unit tests.

okotsopoulos committed:
Include DUOS in TDR service check as non-critical dependency

We call Consent's status endpoint unauthenticated from DuosClient within DuosService with exception handling.
Minor refactoring in StatusService for easier testing.
StatusTest was not a unit test: it made live calls to external services, so I expanded mocking and was able to remove one of our config faults in the process.
Expanded unit tests throughout.

okotsopoulos committed:
Introduce GoogleCredentialsService for obtaining default creds and tokens

This makes it easier to use mocks when  unit test code with which previously relied on GoogleCredentials static calls.
Refactored IamService to call these new service methods.
Added GoogleCredentialsService unit tests.

okotsopoulos committed:
Adjust DuosDaoTest to avoid use of tdrServiceAccountEmail bean

Obtaining this bean involves interaction with GoogleCredentials.

okotsopoulos committed:
Add DuosClient.getApprovedUsers which fetches approved users for a DUOS dataset

- Within user fetch, leverage GoogleCredentialsService to run as TDR SA
- Break out DUOS exception handling into common method - used for both dataset and authorized user fetch
- Add DuosClient unit tests

okotsopoulos committed:
Removed unneeded @VisibleForTesting annotation in standalone DuosDataset record

This record definition was previously embedded within DuosClient

okotsopoulos committed:
Add DuosDao.updateFirecloudGroupLastSyncedDate and unit test

okotsopoulos committed:
Add new endpoint syncDuosDatasetAuthorizedUsers

This allows TDR Admins to force a sync of a TDR-managed Firecloud group associated with the supplied DUOS ID.
If the DUOS ID is not known to TDR, throws a 404.
If we successfully update the Firecloud group's members but fail to record the sync in the DB due to a conflict, throws a 409.  (Note that in this case we do not walk back the sync: in reality, we don't expect this endpoint to be used heavily, and will instead move to automate a batch sync of all TDR-managed Firecloud groups.)

okotsopoulos committed:
Add DuosClient logging for all external calls save for status check

We just log the HTTP method, the URL, and a short description of the authorization used (unauthed, user-authed, or SA-authed -- no PII included)

okotsopoulos committed:
Fix DuosDaoTest.testUpdateFirecloudGroupLastSyncedDate

Previously failed in automation but not locally due to inconsistent truncation of Instant when fetched from DB

okotsopoulos committed:
Add DUOS endpoints retrieveDuosFirecloudGroup(s)

Expanded DuosDao, DuosService, and unit tests.

okotsopoulos committed:
Check for sync_duos_user action on datarepo before enabling sync

Now available in Sam dev: https://github.com/broadinstitute/sam/pull/907

okotsopoulos committed:
Update descriptions on syncAuthorizedUsers endpoint to more accurately reflect behavior

nmalfroy commented on src/main/java/bio/terra/app/controller/DuosApiController.java:
```diff
@@ -0,0 +1,83 @@
+  private final IamService iamService;
+  private final DuosService duosService;
+
+  public DuosApiController(
```
Do we not need an `@Autowired`?

okotsopoulos replied:
We actually don't!  I learned this from [Phil](https://broadworkbench.atlassian.net/wiki/spaces/DSPBLOG/blog/2022/11/03/2579824738/Phil+s+Java+Notes+Dependency+Injection+Phil+Shapiro), it's redundant as of Spring 4.3 :)

nmalfroy commented on src/main/java/bio/terra/service/auth/oauth2/GoogleCredentialsService.java:
```diff
@@ -0,0 +1,56 @@
+import org.springframework.stereotype.Component;
+
+@Component
+public class GoogleCredentialsService {
```
is there a concern that this could get used by different services and you could end up getting state bleed?  That's the problem with doing this in a service and not a static is that you have to account for that now

nmalfroy replied:
I wonder if we could protect against that by not storing scopes separately but just passing the into `getAccessToken`

okotsopoulos replied:
Oooh… great catch.  I agree that this is a real concern!  I do have a preference for keeping this in a service for ease of testing, but I will pass in the scopes instead of setting them on the instance.  We can see if this addresses the issue.

okotsopoulos replied:
I updated this, please let me know if that sufficiently addresses your concern!

nmalfroy commented on src/main/java/bio/terra/service/duos/DuosClient.java:
```diff
@@ -1,31 +1,44 @@
     this.duosConfiguration = duosConfiguration;
     this.restTemplate = restTemplate;
     restTemplate.setRequestFactory(new HttpComponentsClientHttpRequestFactory());
+    this.googleCredentialsService = googleCredentialsService.scopes(DUOS_SCOPES);
```
this is kinda what I meant about sharing state.  If another service also pulls in this dependency then the scopes will be shared and that's no good

nmalfroy commented on src/test/java/bio/terra/app/controller/StatusTest.java:
```diff
@@ -56,36 +71,67 @@ public void testStatusDown() throws Exception {
+        containsSubserviceString(StatusService.DUOS, false, false));
+  }
+
+  private Matcher<String> containsSubserviceString(
```
neat!

nmalfroy commented on src/test/java/bio/terra/service/duos/DuosServiceTest.java:
```diff
@@ -64,9 +157,7 @@ public void testCreateFirecloudGroupWithNamingConflict() {
-    IamConflictException iamConflictEx =
-        new IamConflictException("Group already existed", List.of());
-    doThrow(iamConflictEx).when(iamService).createGroup(FIRECLOUD_GROUP_NAME);
+    doThrow(mock(IamConflictException.class)).when(iamService).createGroup(FIRECLOUD_GROUP_NAME);
```
I'm just curious why change to a mock from an instance?

okotsopoulos replied:
Whoops, missed this question.

I was interested in minimizing definitions and object creation for tests where the internals aren't useful.  In this case, we don't need to check or do anything with the contents of the exception; its class is what dictates the behavior we're testing.

pshapiro4broad replied:
Personally I think I wouldn't mock a non-service class like an exception, unless you needed to mock a method on it. I would write something like
```suggestion
    doThrow(new IamConflictException("", List.of())).when(iamService).createGroup(FIRECLOUD_GROUP_NAME);
```
so it's clear that the contents aren't used

okotsopoulos replied:
That's a good point, I'll do that instead!

nmalfroy reviewed (changes requested):
This looks awesome!  My only real concern is the credentials provider class.  I think there's a state leak in there

snf2ye commented on src/main/java/bio/terra/app/controller/DuosApiController.java:
```diff
@@ -0,0 +1,83 @@
+    return ResponseEntity.ok(duosService.retrieveFirecloudGroup(duosId));
+  }
+
+  @Override
+  public ResponseEntity<DuosFirecloudGroupModel> syncDuosDatasetAuthorizedUsers(String duosId) {
+    iamService.verifyAuthorization(
+        getAuthenticatedInfo(),
+        IamResourceType.DATAREPO,
+        appConfig.getResourceId(),
+        IamAction.SYNC_DUOS_USERS);
+    return ResponseEntity.ok(duosService.syncDuosDatasetAuthorizedUsers(duosId));
```
Just want to sanity check here - this means that the user has access to sync duos users to any dataset, right? Not a specific dataset.

okotsopoulos replied:
Correct! Since DUOS objects are not registered in Sam we can’t get more specific, but we also don’t need to per Jonathan.

snf2ye commented on src/main/java/bio/terra/service/auth/iam/IamRole.java:
```diff
@@ -6,6 +6,7 @@
 
 public enum IamRole {
   ADMIN,
+  MEMBER,
```
Is this the same as the duos admin role that was added to SAM?

okotsopoulos replied:
No, but that’s intentional - this is a role specific to Firecloud groups, which have admins and members.

snf2ye replied:
Is there a good place to add more comments/color to this Role? It's a fairly generic name and I think it could be easily confused with the "user" role.

okotsopoulos replied:
That's good feedback!  I could pursue a slight reorganization akin to what we do for `IamAction`, trying to group by common roles then adding comments for resource-specific role groupings: https://github.com/DataBiosphere/jade-data-repo/blob/a9634d254c9e3caa3d7ce5f60408226b7dfb583b/src/main/java/bio/terra/service/auth/iam/IamAction.java

But this has the potential to go stale.  Minimally I can add a comment disambiguating. Thoughts?

snf2ye replied:
A comment sounds good. 
There isn't any wiggle room on the name of this role, is there?

okotsopoulos replied:
The name itself is not within our control, but we could expand the enum value definition to optionally allow some additional description or context to be specified.  I'll give that a shot.

snf2ye replied:
If that's too much, I'm good with just a comment.

okotsopoulos replied:
I ended up expanding the enum definition to allow for an optional embedded description.

okotsopoulos replied:
Oh, actually… I'm going to undo this!  A Javadoc-style comment is a better fit, and then you can see it in the IntelliJ pop-up window within your references.

okotsopoulos replied:
![Screen Shot 2022-11-23 at 12 44 29 PM](https://user-images.githubusercontent.com/79769153/203614507-e7de3177-e04b-4a50-8bfe-6718041e883d.png)

snf2ye commented on src/main/java/bio/terra/service/status/StatusService.java:
```diff
@@ -19,17 +20,26 @@ public class StatusService {
+  public static final String POSTGRES = "Postgres";
+  public static final String SAM = "Sam";
+  public static final String RBS = "ResourceBufferService";
+  public static final String DUOS = "Consent";
```
Why does this map to "consent" instead of "duos"?

okotsopoulos replied:
Heh, great question…

While we talk about the DUOS integration, the underlying API we're talking to is called Consent.  It's the main API powering DUOS.  But the naming inconsistency didn't sit exactly right with me.  In the case of the status check, I figured we'd want to know that it's Consent specifically which has the issue.

This is moving me towards some renaming: `DuosClient` -> `ConsentClient` but likely keeping most everything else under the DUOS umbrella. That way if we end up talking to other APIs for this integration, they'll have their own clients.  WDYT?

snf2ye replied:
Are we likely to integrate with any other DUOS API? I'd be tempted to go in other direction and name the status check "duos".  Even if we didn't know that it's specifically the consent client, I would think quickly knowing that we should bug the DUOS team would send us in the right direction.  But, I would totally defer to your opinion here. Either way works (or as is!).

okotsopoulos replied:
Good feedback!  I think it's not super likely that we integrate with other DUOS APIs.  If it is, that's more evidence in favor of a DUOS-supplied client.  I'll change this.

snf2ye commented on src/main/java/bio/terra/service/status/StatusService.java:
```diff
@@ -54,19 +65,4 @@ public RepositoryStatusModel getStatus() {
-
-  private RepositoryStatusModelSystems postgresStatus(Boolean isCritical)
-      throws DataAccessException {
-    // Used by Unit test: StatusTest
-    if (configurationService.testInsertFault(ConfigEnum.CRITICAL_SYSTEM_FAULT)) {
-      logger.info(
-          "CRITICAL_SYSTEM_FAULT inserted for test - setting postgres system status to failing");
-      return new RepositoryStatusModelSystems()
-          .ok(false)
-          .critical(true)
-          .message("CRITICAL_SYSTEM_FAULT inserted for test");
-    }
```
Man, oh, man, It can be horrifying to see old code I wrote 😆  - Thanks for fixing this up!
Can we also ditch the fault insertion in getStatus? Don't worry about this if it's too much scope creep.

okotsopoulos replied:
No problem at all, it was fun to get to remove a fault!

When I first tackled this, I wanted to remove this other fault too, which I think would also eliminate our dependency on `configurationService`.  But our existing `StatusTest` uses a real `StatusService`, and what we'd need to do to keep the test is a mocked service.  A bit more than I wanted to do at the time.  I'll take another look and if I don't pursue it now, I've updated this ticket to include it: https://broadworkbench.atlassian.net/browse/DR-2835

okotsopoulos replied:
Took another peek and I will save this for the other ticket!  More extensive test "engoodification" is a better fit there :-)

snf2ye commented on src/main/java/bio/terra/service/duos/DuosClient.java:
```diff
@@ -46,28 +75,66 @@ String getDatasetUrl(String duosId) {
+    logger.info("About to GET {} as TDR SA", url);
+    try {
+      ResponseEntity<DuosDatasetApprovedUsers> approvedUsersCall =
+          restTemplate.exchange(
+              url, HttpMethod.GET, new HttpEntity<>(saHeaders), DuosDatasetApprovedUsers.class);
```
Do you know if there are plans to build out a java client for DUOS so we can avoid doing the direct REST calls?

okotsopoulos replied:
Good question.  Greg filed this ticket on my request but I think it could be awhile before it's implemented.  I could add a link to a code comment as a breadcrumb for later, it's a question that Phil raised on a previous PR too: https://broadworkbench.atlassian.net/browse/DUOS-2191

snf2ye commented on src/main/java/bio/terra/service/duos/DuosService.java:
```diff
@@ -1,21 +1,76 @@
+   * @return TDR-managed DUOS Firecloud group
+   * @throws DuosDatasetNotFoundException if TDR has no record of the DUOS dataset
+   */
+  public DuosFirecloudGroupModel retrieveFirecloudGroup(String duosId) {
+    DuosFirecloudGroupModel firecloudGroup = duosDao.retrieveFirecloudGroupByDuosId(duosId);
+    if (firecloudGroup == null) {
+      throw new DuosDatasetNotFoundException(
+          "DUOS dataset %s has not been registered in TDR".formatted(duosId));
+    }
+    return firecloudGroup;
```
Thoughts on renaming this method to `retrieveFirecloudGroupByDuosId`? I'm finding the existence of multiple versions of `retrieveFirecloudGroup` a little confusing.

okotsopoulos replied:
I appreciate the feedback, I can rename this.

snf2ye commented on src/main/java/bio/terra/app/controller/DuosApiController.java:
```diff
@@ -0,0 +1,83 @@
+  }
+
+  @Override
+  public ResponseEntity<DuosFirecloudGroupModel> retrieveDuosFirecloudGroup(String duosId) {
```
Is the underlying DuosID a UUID? If so, what would you think about switching to passing around the UUID instead of the string?

Historically, we got into a weird state with snapshot & dataset ids where we were having to switch back and forth between string & UUID throughout the code. It's my opinion we should pass around the UUID and only convert to a string directly where used in string format.

However, I could see this being a special case where passing the string makes sense because we never actually use the duosID in our code and only pass it to DUOS (in string format).

And apologies, I think this is baked into some of your previous changes, I should I have made this comment earlier in case we do want to change it.

okotsopoulos replied:
I am in complete agreement with you, re: UUID conversion.  In this case, duosId is not a UUID - it's a true string of the format `DUOS-<nnnnnn>` like `DUOS-123456`.  I give example values in the OpenAPI spec but I don't think I do in the code itself.

Would it help if I updated these parameter descriptions to clarify the format?  My concern with that is that we don't enforce this format ourselves: we delegate it to DUOS, so that if they change their formatting requirements we're not out of sync.

snf2ye replied:
Ah, sorry I missed that - The example in the open API spec is sufficient.

snf2ye commented on src/main/java/bio/terra/service/duos/SystemStatus.java:
```diff
@@ -0,0 +1,8 @@
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import java.util.Map;
+
+@JsonIgnoreProperties(ignoreUnknown = true)
```
Is this tag always needed for records?

okotsopoulos replied:
Good question, it isn't required but I like it in cases like this one.  I.E. we would really prefer a client provided by DUOS team, but because we're writing our own based off of the existing responses given by their API.  This avoids deserialization errors if they ultimately expand their response bodies, we'll just ignore any values we don't care about.

okotsopoulos replied:
(It's also nice if you know you only care about a small subset of existing fields!)

snf2ye replied:
Ah gotcha, thanks for explaining!

okotsopoulos committed:
Ensure that GoogleCredentialsService does not leak state among callers

Instead, make callers pass in requested scopes.

okotsopoulos committed:
Register DUOS status check as DUOS rather than Consent

Consent is the name of the API we call, but flagging an issue with DUOS is more likely to point us in the right direction, i.e. towards the DUOS team and Slack channel of the same name.

okotsopoulos committed:
Update IamRole.MEMBER with description clarifying its usage in Firecloud groups

It's a generic enough role name that it could be confused with IamRole.USER, which we reference frequently.

okotsopoulos committed:
Remove IamRole embedded description in favor of Javadoc comment for MEMBER

You then see the Javadoc comment when hovering on the reference in IntelliJ.

pshapiro4broad commented on src/main/java/bio/terra/app/controller/DuosApiController.java:
```diff
@@ -0,0 +1,83 @@
+@Api(tags = {"duos"})
+public class DuosApiController implements DuosApi {
+
+  private final Logger logger = LoggerFactory.getLogger(DuosApiController.class);
```
This appears to be unused
```suggestion
```

pshapiro4broad commented on src/main/java/bio/terra/app/controller/DuosApiController.java:
```diff
@@ -0,0 +1,83 @@
+import org.springframework.stereotype.Controller;
+
+@Controller
+@Api(tags = {"duos"})
```
Is this necessary? The API is scanned already (so `@Api` isn't needed) and I think each method in the controller already has the operation tags `duos` and `repository` (you can see this in the `@Operation` annotation on the methods in `DuosApi`).

If it is necessary, I wonder why it's different from the tags specified in the openapi yaml, which has both `duos` and `repository` tags.

okotsopoulos replied:
Good question.  We've added this annotation for our controllers throughout but it may not be needed, I'll test out what happens when we remove it.

pshapiro4broad commented on src/main/java/bio/terra/app/controller/DuosApiController.java:
```diff
@@ -0,0 +1,83 @@
+  }
+
+  @Override
+  public Optional<HttpServletRequest> getRequest() {
+    return Optional.ofNullable(request);
+  }
+
+  @Override
+  public Optional<ObjectMapper> getObjectMapper() {
```
I wonder if these overrides are needed. We don't implement them in catalog and the APIs seem to work fine. I think they're only used if you don't override any controller methods (e.g., you rely on the default implementations in the controller base class). So they should be unused in TDR.

okotsopoulos replied:
Same here, I'll test out their removal.

okotsopoulos replied:
These are the default implementations in the controller base class:

```
    default Optional<ObjectMapper> getObjectMapper(){
        return Optional.empty();
    }

    default Optional<HttpServletRequest> getRequest(){
        return Optional.empty();
    }
```

I'm not 100% sure where the defaults come from or how to override at a higher level, but I'll elect to leave these controller-level overrides.

pshapiro4broad replied:
Yeah, but the defaults are only called from the `default` methods in the swagger generated controller class (and all the default methods are overridden by our controllers). We don't call them in our code and nothing else in the generated code refers to them, as far as I know. You can use IntelliJ's Find Usages to verify that nothing else is calling these methods.

okotsopoulos replied:
Ahhhh!  That clicks for me now.  Appreciate the further explanation.

pshapiro4broad commented on src/main/java/bio/terra/app/controller/DuosApiController.java:
```diff
@@ -0,0 +1,83 @@
+
+  @Override
+  public ResponseEntity<DuosFirecloudGroupModel> retrieveDuosFirecloudGroup(String duosId) {
+    return ResponseEntity.ok(duosService.retrieveFirecloudGroup(duosId));
```
Is there no auth check here? If not, does it need to be an auth'd API? My guess is that it's more trouble than it's worth to make this a non-auth API, but it does seem odd to me to require auth but not use it. Maybe this is not uncommon in TDR?

okotsopoulos replied:
Ah, I'm glad you mentioned this because it's not that common a pattern in TDR.  We want to ensure that the user is logged in but don't have a specific permission to check, which is why I didn't make it an unauthed endpoint… the root object here is a DUOS dataset's metadata which is accessible to logged in users in DUOS.  DUOS datasets aren't resources in Sam.

pshapiro4broad commented on src/main/java/bio/terra/service/auth/iam/IamService.java:
```diff
@@ -356,19 +357,15 @@ public UserStatusInfo getUserInfo(AuthenticatedUserRequest userReq) {
-      throw new GoogleResourceException("Could not generate Google credentials", e);
-    }
-    return callProvider(() -> iamProvider.registerUser(getAccessToken(impersonatedCredentials)));
+    ImpersonatedCredentials impersonatedCredentials =
+        ImpersonatedCredentials.create(
+            googleCredentialsService.getApplicationDefault(),
+            serviceAccountEmail,
+            null,
+            SCOPES,
+            (int) TOKEN_LENGTH.toSeconds());
+    String accessToken = googleCredentialsService.getAccessToken(impersonatedCredentials, SCOPES);
```
Could this logic be moved into `GoogleCredentialsService`? The API could look something like
```
  String accessToken = googleCredentialsService.getImpersonatedAccessToken(serviceAccountEmail, SCOPES)
```

okotsopoulos replied:
I think it could be.  I was trying to keep the scope of my refactoring narrow but I'll take a closer look.

pshapiro4broad commented on src/main/java/bio/terra/service/auth/oauth2/GoogleCredentialsService.java:
```diff
@@ -0,0 +1,45 @@
+  /**
+   * @return valid oauth token for credentials with scopes specified when required
+   */
+  public String getAccessToken(GoogleCredentials credentials, List<String> scopes) {
```
It looks like there is similar code in `GoogleBillingService.accessClient()`, `GoogleProjectService.iamService()` and `ResourceBufferServiceConfiguration.getAccessToken()`. Would it be possible / worthwhile to refactor these usages too?

okotsopoulos replied:
Similar comment on refactoring scope, but if not now, when? :) I will look at these.

pshapiro4broad commented on src/main/java/bio/terra/service/duos/DuosClient.java:
```diff
@@ -1,31 +1,44 @@
 public class DuosClient {
 
+  private static final Logger logger = LoggerFactory.getLogger(DuosClient.class);
+  @VisibleForTesting static final List<String> SCOPES = List.of("email", "profile");
```
IMO for a static final (and immutable) member there's no need to annotate as `@VisibleForTesting`, as there's not much danger of exposing internals of a class.

pshapiro4broad commented on src/main/java/bio/terra/service/duos/DuosClient.java:
```diff
@@ -46,28 +75,66 @@ String getDatasetUrl(String duosId) {
     HttpHeaders authedHeaders = getHttpHeaders();
     authedHeaders.setBearerAuth(userRequest.getToken());
+    String url = getDatasetUrl(duosId);
+    logger.info("About to GET {} as authenticated user", url);
```
Did you mean to leave this `info()` in place? It looks like a debugging message.

okotsopoulos replied:
Good point, I meant to leave it in but it's a better fit for debug logs. I will downgrade.

pshapiro4broad commented on src/main/java/bio/terra/service/duos/DuosClient.java:
```diff
@@ -46,28 +75,66 @@ String getDatasetUrl(String duosId) {
+  @VisibleForTesting
+  static ErrorReportException convertDuosExToDataRepoEx(
+      HttpStatusCodeException duosEx, String duosId) {
+    switch (duosEx.getStatusCode()) {
+      case BAD_REQUEST:
+        {
+          return new DuosDatasetBadRequestException(
+              "DUOS dataset identifier %s is malformed".formatted(duosId), duosEx);
+        }
+      case NOT_FOUND:
+        {
+          return new DuosDatasetNotFoundException(
+              "Could not find DUOS dataset for identifier %s".formatted(duosId), duosEx);
+        }
+      default:
+        {
+          return new DuosInternalServerErrorException("Unexpected error from DUOS", duosEx);
+        }
```
I think you can replace this with a switch expression
```suggestion
    return switch (duosEx.getStatusCode()) {
      case BAD_REQUEST -> new DuosDatasetBadRequestException(
          "DUOS dataset identifier %s is malformed".formatted(duosId), duosEx);
      case NOT_FOUND -> new DuosDatasetNotFoundException(
          "Could not find DUOS dataset for identifier %s".formatted(duosId), duosEx);
      default -> new DuosInternalServerErrorException("Unexpected error from DUOS", duosEx);
    };
```

okotsopoulos replied:
Ah, thanks! I was struggling with this for some reason, will check it out.

pshapiro4broad replied:
I think IntelliJ can do this for you, if the right settings are enabled.

pshapiro4broad commented on src/main/java/bio/terra/service/duos/DuosClient.java:
```diff
@@ -46,28 +75,66 @@ String getDatasetUrl(String duosId) {
+    String url = getApprovedUsersUrl(duosId);
+    logger.info("About to GET {} as TDR SA", url);
+    try {
+      ResponseEntity<DuosDatasetApprovedUsers> approvedUsersCall =
```
Totally up to you, but I'd use `var` for each of these `ResponseEntity` types. I don't find that having the type declared makes it clearer, but it's a matter of taste.
```suggestion
      var approvedUsersCall =
```

pshapiro4broad commented on src/main/java/bio/terra/service/duos/DuosClient.java:
```diff
@@ -46,28 +75,66 @@ String getDatasetUrl(String duosId) {
+  }
+
+  @VisibleForTesting
+  static ErrorReportException convertDuosExToDataRepoEx(
```
Personally I'm not a fan of abbreviations like `Ex`. Writing out the full name both times here is wordy though. Since it's a private method (or it would be, if it wasn't visible for testing) using a shorter name like `convertToDataRepoException()` would be fine with me.

okotsopoulos replied:
Good feedback.  I was following a naming pattern used for Sam exception conversion, but I like your suggestion.

pshapiro4broad commented on src/main/java/bio/terra/service/duos/DuosDao.java:
```diff
@@ -75,6 +71,14 @@ public UUID insertFirecloudGroup(DuosFirecloudGroupModel created) {
+      isolation = Isolation.SERIALIZABLE,
+      readOnly = true)
+  public List<DuosFirecloudGroupModel> retrieveFirecloudGroups() {
+    return jdbcTemplate.query(duosFirecloudGroupQuery, new DuosFirecloudGroupMapper());
```
It occurs to me that this mapper is stateless so you could create it once and store it as a static final in the class, something like
```
    private static final DuosFirecloudGroupMapper MAPPER = new DuosFirecloudGroupMapper();
```

pshapiro4broad commented on src/main/java/bio/terra/service/duos/DuosDao.java:
```diff
@@ -103,6 +107,23 @@ public DuosFirecloudGroupModel retrieveFirecloudGroupByDuosId(String duosId) {
+            SET last_synced_date = :last_synced_date
+            WHERE id = :id
+            """;
+    MapSqlParameterSource params =
+        new MapSqlParameterSource()
+            .addValue("id", id)
+            .addValue("last_synced_date", Timestamp.from(lastSyncedDate));
```
Now that we have `Map.of()` I don't use `MapSqlParameterSource` anymore. Maybe I'm missing some benefit that `MapSqlParameterSource` provides?
```suggestion
    var params = Map.of("id", id, "last_synced_date", Timestamp.from(lastSyncedDate));
```

okotsopoulos replied:
Good to know!  Personal preference but I like the clear grouping of parameters to values that `MapSqlParameterSource` provides.

pshapiro4broad commented on src/main/java/bio/terra/service/duos/DuosDao.java:
```diff
@@ -103,6 +107,23 @@ public DuosFirecloudGroupModel retrieveFirecloudGroupByDuosId(String duosId) {
+            .addValue("id", id)
+            .addValue("last_synced_date", Timestamp.from(lastSyncedDate));
+    int rowsAffected = jdbcTemplate.update(sql, params);
+    return rowsAffected > 0;
```
This return value isn't used in the prod code, is there a reason to return it?

pshapiro4broad commented on src/test/java/bio/terra/app/controller/StatusTest.java:
```diff
@@ -56,36 +71,67 @@ public void testStatusDown() throws Exception {
-        upResponseBody,
+        "/Status response should indicate that the system is up.",
+        responseBody,
         startsWith("{\"ok\":true"));
```
Instead of string matching on JSON, I think it would be better to either use `JSONAssert` or `JsonPath`. Both are part of spring boot test so TDR already has them as a test dependency. For example, using `JsonPath` you could write:

```
    assertThat(
        "/Status response should indicate that the system is up.",
        JsonPath.read(responseBody, "$.ok"),
        is("true"));
    assertThat(
        "/Status response should indicate that postgres is up",
        JsonPath.read(responseBody, "$.Postgres.ok"),
        is(true));
```

pshapiro4broad replied:
`JsonPath` is built into MvcResult so you could also write
```
    mvc.perform(get("/status")).andExpect(status().isOk())
        .andExpect(jsonPath("$.ok", is("true")))
        .andExpect(jsonPath("$.Postgres.ok", is("true")));
```

okotsopoulos replied:
Thanks!  This was an older test, it will be nice to remove the string matching (either within this PR or as part of the standalone ticket for expanding our status endpoint: https://broadworkbench.atlassian.net/browse/DR-2835)

pshapiro4broad reviewed (commented):
Here's my comments so far, am about halfway through.

pshapiro4broad commented on src/main/java/bio/terra/service/duos/DuosDao.java:
```diff
@@ -103,6 +107,23 @@ public DuosFirecloudGroupModel retrieveFirecloudGroupByDuosId(String duosId) {
   }
 
+  @Transactional(propagation = Propagation.REQUIRED, isolation = Isolation.SERIALIZABLE)
+  public boolean updateFirecloudGroupLastSyncedDate(UUID id, Instant lastSyncedDate) {
```
When is `lastSyncedDate` not `Instant.now()`? Unless there's need to make this some other time, I'd remove this argument. I think you could have postgres create a 'now' timestamp for you too.

pshapiro4broad commented on src/main/java/bio/terra/service/duos/DuosService.java:
```diff
@@ -54,4 +109,79 @@ String constructFirecloudGroupName(String duosId) {
+   *     of approved users
+   * @return DUOS Firecloud group reflecting the last successful sync of its members
+   */
+  @VisibleForTesting
```
This PR uses a lot of `VisibleForTesting`; I think in general it's better to use this sparingly. If there's an API that would otherwise be `private`, then it can be tested via other APIs that are part of the class's visible interface. Testing non-public APIs means that the tests are more fragile and may need to be updated when a private API is changed.

okotsopoulos replied:
This was helpful for me to think about.  Going back to this class in particular, I found that my tests on the public methods already covered much of what I wanted to test in those made `VisibleForTesting`, so exposing them in that way was overkill.

pshapiro4broad commented on src/main/java/bio/terra/service/duos/DuosService.java:
```diff
@@ -54,4 +109,79 @@ String constructFirecloudGroupName(String duosId) {
+              "were updated, but could not update DUOS Firecloud Group record",
+              id.toString(),
+              "with last_synced_date",
+              lastSyncedDate.toString());
```
This use of `String.join()` seems unusual to me. Normally `join()` is called on a collection or list of variables. In this case it looks like another way of writing a string concatenation. Using `+` has the benefit that you don't need to all `toString()`, and (to me) seems like a more idiomatic way to write this in Java. Another way to write this would be to use `formatted()`:
```suggestion
      String msg =
          ("Members of Firecloud group %s were updated, but could not update DUOS Firecloud " +
              "Group record %s with last_synced_date %s")
              .formatted(firecloudGroup.getFirecloudGroupName(), id, lastSyncedDate);
```

pshapiro4broad commented on src/main/java/bio/terra/service/status/StatusService.java:
```diff
@@ -42,9 +52,10 @@ public RepositoryStatusModel getStatus() {
-    statusModel.putSystemsItem("Postgres", postgresStatus(true));
-    statusModel.putSystemsItem("Sam", iamProviderInterface.samStatus().critical(true));
-    statusModel.putSystemsItem("ResourceBufferService", bufferService.status());
+    statusModel.putSystemsItem(POSTGRES, datasetDao.statusCheck().critical(true));
```
This is pretty close to what catalog does, with the difference that catalog has each subsystem add itself to the list of services. So this class doesn't need to change when a new service is added. If you're interested, the code is: https://github.com/DataBiosphere/terra-data-catalog/blob/main/service/src/main/java/bio/terra/catalog/service/BaseStatusService.java

okotsopoulos replied:
That's great! I added it to our status check improvement ticket: https://broadworkbench.atlassian.net/browse/DR-2835

pshapiro4broad commented on src/test/java/bio/terra/service/auth/iam/sam/SamIamTest.java:
```diff
@@ -642,14 +642,13 @@ public void testCreateGroup() throws ApiException, InterruptedException {
-    verify(samGroupApi, times(1)).getGroup(groupName);
+        equalTo(groupEmail));
+    verify(samGroupApi).postGroup(groupName);
+    verify(samGroupApi).getGroup(groupName);
```
I followed up in slack, but for the record, if strict stubs are enabled then this `verify()` is redundant since unused mocks are errors in strict mode. The problem is the current way strict mode is enabled in TDR doesn't seem to work. Using
```
  @Rule public MockitoRule mockito = MockitoJUnit.rule().strictness(Strictness.STRICT_STUBS);
```
does seem to work, and once it's added, you no longer need `@RunWith(MockitoJUnitRunner.StrictStubs.class)` on the test class.

pshapiro4broad commented on src/test/java/bio/terra/service/auth/oauth2/GoogleCredentialsServiceTest.java:
```diff
@@ -0,0 +1,81 @@
+
+  private GoogleCredentialsService service;
+
+  private static final List<String> SCOPES = List.of("openid", "email", "profile");
```
Is the content of `SCOPES` important for this test? If it's not, can I suggest using more obvious dummy values here? That makes it clearer to a future maintainer that these values don't need to be changed when other values (like those in `IamService.SCOPES`) change.

okotsopoulos replied:
That's a good point! I changed these values.

pshapiro4broad commented on src/main/java/bio/terra/service/auth/iam/IamService.java:
```diff
@@ -62,15 +60,18 @@ public class IamService {
   private final ConfigurationService configurationService;
```
It looks like this field can be removed, as it's unused except in the constructor.

pshapiro4broad commented on src/test/java/bio/terra/service/duos/DuosClientTest.java:
```diff
@@ -106,9 +147,85 @@ public void testGetDatasetThrowsWhenIdMalformed() {
+  public void testGetApprovedUsers() {
+    var expectedBody =
+        new DuosDatasetApprovedUsers(
+            List.of("a", "b", "c").stream().map(DuosDatasetApprovedUser::new).toList());
```
You can create a stream directly from a list of objets
```suggestion
            Stream.of("a", "b", "c").map(DuosDatasetApprovedUser::new).toList());
```

okotsopoulos committed:
Removed unused logger and @Api annotation in DuosApiController

The @Api annotation is unnecessary despite it being one that we use for our controllers: the overridden autogenerated API methods are tagged appropriately.

pshapiro4broad commented on src/test/java/bio/terra/service/duos/DuosClientTest.java:
```diff
@@ -106,9 +147,85 @@ public void testGetDatasetThrowsWhenIdMalformed() {
+  }
+
+  // Utilities for examining HttpEntity parameters
+  private List<String> getAuthorizations(HttpEntity httpEntity) {
```
`HttpEntity` is a generic type and it's generally considered bad form to use a generic type without a type argument, and the java compiler will generate a warning (as does IntelliJ). It's usually easy to fix by either using a wildcard type (`<?>`) or by making the class or method that uses it generic. In this case you can write
```suggestion
  private <T> List<String> getAuthorizations(HttpEntity<T> httpEntity) {
```

Note that these helper methods could all be made `static`. If you do that, you'll need to change the method references from `this::isUnauthenticated` to `DuosClientTest::isUnauthenticated`. My thinking is that making a method static tells the reader that the method doesn't use any class state. In a test class it's less important but IMO it's still a nice rule to follow.

okotsopoulos replied:
Ahh, I missed those warnings, thanks for pointing them out.  I agree with you re: what's implied by making these static methods, and made those changes as well.

pshapiro4broad commented on src/test/java/bio/terra/service/duos/DuosDaoTest.java:
```diff
@@ -33,21 +37,24 @@
 
   @Before
   public void before() {
+    duosDao = new DuosDao(jdbcTemplate, TDR_SERVICE_ACCOUNT_EMAIL);
```
If you wanted to work around this need to construct the Dao class by hand, you could convert the service account email to a record type (instead of a qualified String) and then mock it in this class:
```
  @MockBean private ServiceAccountEmail serviceAccountEmail;
```
Using a record type would let you get rid of all the `@Qualified` attributes too, if that's something you've been thinking about doing

pshapiro4broad commented on src/test/java/bio/terra/service/duos/DuosDaoTest.java:
```diff
@@ -83,6 +92,10 @@ public void testInsertThenRetrieveFirecloudGroup() {
     verifyRetrievedFirecloudGroupContents(retrievedById);
+
+    List<DuosFirecloudGroupModel> retrieved = duosDao.retrieveFirecloudGroups();
+    assertThat(retrieved.size(), equalTo(1));
```
There are matchers for collection length
```suggestion
    assertThat(retrieved, hasSize(1));
```

pshapiro4broad commented on src/test/java/bio/terra/service/duos/DuosDaoTest.java:
```diff
@@ -66,6 +73,8 @@ public void after() {
   public void testRetrieveFirecloudGroupBeforeInsert() {
     DuosFirecloudGroupModel retrieveBeforeInsert = duosDao.retrieveFirecloudGroupByDuosId(duosId);
     assertNull(retrieveBeforeInsert);
+    List<DuosFirecloudGroupModel> retrieveAllBeforeInsert = duosDao.retrieveFirecloudGroups();
```
This seems like a good place to use `var`
```suggestion
    var retrieveAllBeforeInsert = duosDao.retrieveFirecloudGroups();
```

pshapiro4broad commented on src/test/java/bio/terra/service/duos/DuosDaoTest.java:
```diff
@@ -107,13 +120,37 @@ public void testInsertFirecloudGroupThrowsOnDuplicateDuosId() {
+  @Test
+  public void testUpdateFirecloudGroupLastSyncedDate() {
+    duosFirecloudGroupId = duosDao.insertFirecloudGroup(toInsert);
+    Instant lastSyncedDate = Instant.parse("2022-11-17T00:00:00.00Z");
```
Is there a reason to not use `now()` here?

okotsopoulos replied:
Yes there is!

https://github.com/DataBiosphere/jade-data-repo/pull/1376/commits/e9f09ed0fc69c1e2b4a892e34b9e8df1c613d81e

But that's not obvious looking at the code. I can add a comment.

pshapiro4broad replied:
It looks like it's needed because the database truncates the timestamp (so it's really a date stamp, not a time stamp). If so, I think it would be better to truncate in the test instead of hardcoding a value. I think `Instant.truncatedTo(ChronoUnit.DAYS)` could do this.

Looking at the code, I don't see any use of passing in an explicit time to `updateFirecloudGroupLastSyncedDate()` except in this test. IMO it would be better to remove this argument and use some other way to validate the updated date. One way would be to save the value of "now" before running the sync, and assert that the sync time is after the saved time.

pshapiro4broad reviewed (approved):
Many comments, but overall this looks OK to me. It does make me wonder about the amount of code being added to TDR that could just as well live inside duos (as it doesn't rely on any TDR state that I can see, other than SAM permissions).

okotsopoulos committed:
Merge branch 'develop' into okotsopo/DR-2794-sync-duos-group-contents

okotsopoulos committed:
Mocking exceptions in tests is overkill, we don't need to mock their methods

okotsopoulos committed:
Remove unneeded DuosApiController getters for request, object mapper

We see these in all of our controllers but they are only used in the default implementations of the api methods, which we override.
We can easily add them back in if we prefer that to be an explicit choice.

okotsopoulos committed:
Improvements to DuosClient and unit tests

- Downgrade logging to debug
- Concise switch statement rather than returning within when converting DUOS exceptions

okotsopoulos committed:
Downgrade most DuosService helper methods to private

Ensure unit test coverage for public methods also tests the expected outcomes from these calls
This loosens the coupling between internal implementation details and unit tests

okotsopoulos committed:
Favor obvious dummy scope names in GoogleCredentialsServiceTest

okotsopoulos committed:
Remove unused instance variable in IamService

okotsopoulos committed:
PR feedback for DuosDao and tests

okotsopoulos committed:
Merge branch 'develop' into okotsopo/DR-2793-sync-duos-group-contents

okotsopoulos commented:
Note: this PR still has "changes requested" from Muscles' review, but I addressed the state leak concern.  So I feel okay about forcing a merge with him OOO.
